storage/gcsupload/gcsupload_test.go

```diff
@@ -0,0 +1,69 @@
+// Copyright 2017 Google Inc. All rights reserved.
+// Use of this source code is governed by the Apache 2.0
+// license that can be found in the LICENSE file.
+
+package main
+
+import (
+	"bytes"
+	"crypto/md5"
+	"strings"
+	"testing"
+
+	"golang.org/x/net/context"
+
+	"cloud.google.com/go/storage"
+
+	"github.com/GoogleCloudPlatform/golang-samples/internal/testutil"
+)
+
+func TestUpload(t *testing.T) {
+	tc := testutil.SystemTest(t)
+
+	ctx := context.Background()
+	bh, cleanupBucket, err := withBucketCleanup(t, ctx, tc.ProjectID, "gcs-upload-test")
+	if err != nil {
+		t.Fatalf("setting bucket: %v", err)
+	}
+	defer cleanupBucket()
+
+	input := strings.Repeat("GCS test\n", 30)
+	r := strings.NewReader(input)
+
+	name := "atest.txt"
+	obj, objAttrs, err := upload(ctx, r, tc.ProjectID, bh.Name, name, true)
+	if err != nil {
+		t.Fatalf("expected to successfully upload: %v", err)
+	}
+	if objAttrs == nil {
+		t.Fatal("expected back a non-nil object")
+	}
+	defer obj.Delete(ctx)
+
+	if g, w := objAttrs.Name, name; g != w {
+		t.Errorf("name: got=%q want=%q", g, w)
+	}
+	if g, w := objAttrs.Size, int64(len(input)); g != w {
+		t.Errorf("size: got=%d want=%d", g, w)
+	}
+	h := md5.New()
+	h.Write([]byte(input))
+	if g, w := objAttrs.MD5, h.Sum(nil); !bytes.Equal(g, w) {
+		t.Errorf("md5: got=%x want=%x", g, w)
+	}
+}
+
+func withBucketCleanup(t *testing.T, ctx context.Context, projectID, bucket string) (*storage.BucketAttrs, func() error, error) {
```

Comment:
Looks good, you may want some more code that cleans up the bucket. Maybe this should live in a common test helper package...

https://github.com/GoogleCloudPlatform/golang-samples/blob/master/storage/objects/main_test.go#L176

And see related PR:
https://github.com/GoogleCloudPlatform/golang-samples/pull/230

(these changes haven't been made to the storage/buckets package)